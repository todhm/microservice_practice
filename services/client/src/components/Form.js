import React from 'react';

export default ({formData,formType, handleFormChange,handleUsersFormSubmit})=>(
    <div>
        <h1 className="title is-1">{formType}</h1>
        <hr/><br/>
        <form onSubmit={(event)=>handleUsersFormSubmit(event)}>
            {formType === 'Register' &&
            <div className="field">
                <input
                name="username"
                className="input is-medium"
                type="text"
                placeholder="Enter a username"
                required 
                value={formData.username}
                onChange={handleFormChange}
                />
            </div>
            }
            <div className="field">
                <input
                    name="email"
                    className="input is-medium"
                    type="email"
                    placeholder="Enter an email address"
                    required 
                    value={formData.email}
                    onChange={handleFormChange}
                />
            </div>
            <div className="field">
                <input
                    name="password"
                    className="input is-medium"
                    type="password"
                    placeholder="Enter a password"
                    required 
                    value={formData.password}
                    onChange={handleFormChange}
                />
            </div>
            <input 
            type="submit"
            className="button is-primary is-medium is-fullwidth"
            value="Submit"
            />
        </form>
    </div>
)