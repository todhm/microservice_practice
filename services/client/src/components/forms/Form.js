import axios from 'axios';
import React,{Component} from 'react';
import {Redirect} from 'react-router-dom';
import {registerFormRules, loginFormRules} from './form-rules.js';
import FormErrors from './FormErrors';
export default class Form extends Component{
    state = {
        formData:{
            username:'',
            email:'',
            password: ''
        },
        registerFormRules:registerFormRules, 
        loginFormRules:loginFormRules,
        valid:false,
    }


    componentDidMount=()=>{
        this.clearForm(); 
        this.validateForm(); 
    }

    componentWillReceiveProps=(nextProps)=>{
        if(this.props.formType !== nextProps.formType){
            this.clearForm(); 
            this.validateForm(); 
        }
    }

    clearForm=()=>{
        this.setState({
            formData:{username:'',email:'', password:''}
        });
    };

    handleFormChange=(event)=>{
        const obj = this.state.formData; 
        obj[event.target.name] = event.target.value; 
        this.setState(obj);
        this.validateForm();
    }

    handleUserFormSubmit=(event)=>{
        event.preventDefault(); 
        const formType = this.props.formType
        const data = {
            email: this.state.formData.email, 
            password: this.state.formData.password
        };
        if(formType === 'Register'){
            data.username = this.state.formData.username; 
        }
        const loweredFormType = formType.toLowerCase();
        const url = `${process.env.REACT_APP_USERS_SERVICE_URL}/auth/${loweredFormType}`
        axios.post(url,data)
        .then((res)=>{
            this.clearForm(); 
            this.props.loginUser(res.data.auth_token);
        })
        .catch((err)=>{
            if(formType === 'Login'){
                this.props.createMessage('User does not exists.','danger');

            }
            if(formType==='Register'){
                this.props.createMessage('That user already exists.','danger');
            }
        });
    };

    validateForm=()=>{
        const formData = this.state.formData; 
        this.resetRules(); 
        if(this.props.formType==='Register'){
            const formRules = this.state.registerFormRules; 
            if(formData.username.length>5) formRules[0].valid=true; 
            if(formData.email.length>5) formRules[1].valid=true; 
            if(this.validateEmail(formData.email)) formRules[2].valid=true; 
            if(formData.password.length > 10) formRules[3].valid=true; 
            this.setState({registerFormRules:formRules})
            if(this.allTrue()){
                this.setState({valid:true});
            }
        }
        if(this.props.formType==='Login'){
            const formRules = this.state.loginFormRules; 
            if(formData.email.length > 0)formRules[0].valid=true; 
            if(formData.password.length > 0) formRules[1].valid = true; 
            this.setState({loginFormRules: formRules})
            if(this.allTrue()) this.setState({valid:true});
        }
    }
    validateEmail=(email)=>{
        var re = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
        return re.test(email);

    }
    allTrue=()=>{
        let formRules = loginFormRules; 
        if(this.props.formType ==='Register'){
            formRules = registerFormRules; 
        }
        for(const rule of formRules){
            if(!rule.valid) return false; 
        }
        return true; 
    };
    resetRules=()=>{
        const registerFormRules = this.state.registerFormRules;
        for(const rule of registerFormRules){
            rule.valid = false; 
        }
        this.setState({registerFormRules:registerFormRules});
        const loginFormRules = this.state.loginFormRules; 
        for(const rule of loginFormRules){
            rule.valid = false; 
        }
        this.setState({loginFormRules: loginFormRules,valid:false});
    }

    render(){
        const{ isAuthenticated ,formType} = this.props; 
        const{formData, valid,loginFormRules, registerFormRules} = this.state; 
        if(isAuthenticated){
            return <Redirect to = '/'/>
        }
        let formRules = loginFormRules;
        if(formType==='Register'){
            formRules = registerFormRules; 
        }
        return(
        <div>
            <h1 className="title is-1">{formType}</h1>
            <hr/><br/>
            <FormErrors 
                formType={formType}
                formRules={formRules}
            />
            <form onSubmit={this.handleUserFormSubmit}>
                {formType === 'Register' &&
                <div className="field">
                    <input
                    name="username"
                    className="input is-medium"
                    type="text"
                    placeholder="Enter a username"
                    required 
                    value={formData.username}
                    onChange={this.handleFormChange}
                    />
                </div>
                }
                <div className="field">
                    <input
                        name="email"
                        className="input is-medium"
                        type="email"
                        placeholder="Enter an email address"
                        required 
                        value={formData.email}
                        onChange={this.handleFormChange}
                    />
                </div>
                <div className="field">
                    <input
                        name="password"
                        className="input is-medium"
                        type="password"
                        placeholder="Enter a password"
                        required 
                        value={formData.password}
                        onChange={this.handleFormChange}
                    />
                </div>
                <input 
                type="submit"
                className="button is-primary is-medium is-fullwidth"
                value="Submit"
                disabled={!valid}
                />
            </form>
        </div>
        )
    }



}

