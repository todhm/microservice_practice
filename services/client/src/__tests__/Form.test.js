import React from 'react';
import {shallow} from 'enzyme';
import renderer from 'react-test-renderer';

import Form from '../components/forms/Form';

const testData=[
    {
        formType:'Register',
        formData :{
            username: '',
            email: '',
            password: '',
        },
        handleUserFormSubmit:jest.fn(),
        handleFormChange:jest.fn(),
        isAuthenticated:false,
    },
    {
        formType:'Login',
        formData :{
            email: '',
            password: '',
        },
        handleUserFormSubmit:jest.fn(),
        handleFormChange:jest.fn(),
        isAuthenticated:false,
    },
]
const formData={
    username: '',
    email: '',
    password: '',
}

describe('When not authenticated',()=>{

    testData.forEach((el)=>{
        const component = <Form {...el}/>;
        it(`${el.formType} Form renders properly`,()=>{
            const wrapper = shallow(component);
            const h1 = wrapper.find('h1');
            expect(h1.length).toBe(1);
            expect(h1.get(0).props.children).toBe(el.formType);
            const formGroup = wrapper.find('.field');
            expect(formGroup.length).toBe(Object.keys(el.formData).length);
            expect(formGroup.get(0).props.children.props.name).toBe(Object.keys(el.formData)[0]);
            expect(formGroup.get(0).props.children.props.value).toBe('');
            
        })
        it(`${el.formType} Form renders a snapshot properly`,()=>{
            const tree = renderer.create(component).toJSON();
            expect(tree).toMatchSnapshot();
        
        })

        it(`${el.formType} Form submits the form properly`,()=>{
            const wrapper = shallow(component);
            wrapper.instance().handleUserFormSubmit = el.handleUserFormSubmit; 
            wrapper.instance().validateForm = jest.fn(); 
            wrapper.update(); 
           const input = wrapper.find('input[type="email"]');
           expect(wrapper.instance().handleUserFormSubmit).toHaveBeenCalledTimes(0);
           input.simulate(
               'change', {target:{name:'email',value:'test@test.com'}}
           )
           wrapper.find("form").simulate('submit',el.formData)
           expect(wrapper.instance().handleUserFormSubmit).toHaveBeenCalledWith(el.formData);
           expect(wrapper.instance().handleUserFormSubmit).toHaveBeenCalledTimes(1);
           expect(wrapper.instance().validateForm).toHaveBeenCalledTimes(1);
        })

        it(`${el.formType} Form should be disabled by default`,()=>{
            const wrapper = shallow(component);
            const input = wrapper.find('input[type="submit"]');
            expect(input.get(0).props.disabled).toEqual(true);

        })
    })    
});

describe('When authenticated',()=>{
    testData.forEach((el)=>{
        const component = <Form 
            formType={el.formType}
            formData={el.formData}
            isAuthenticated={true}
            />;
        it(`${el.formType} redirects properly`,()=>{
            const wrapper = shallow(component);
            expect(wrapper.find('Redirect')).toHaveLength(1);
        })

    })
})
