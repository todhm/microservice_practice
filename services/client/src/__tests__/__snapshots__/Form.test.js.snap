// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`{el.formType}\` Form renders a snapshot properly 1`] = `
<div>
  <h1
    className="title is-1"
  >
    Register
  </h1>
  <hr />
  <br />
  <form
    onSubmit={[Function]}
  >
    <div
      className="field"
    >
      <input
        className="input is-medium"
        name="username"
        onChange={undefined}
        placeholder="Enter a username"
        required={true}
        type="text"
        value=""
      />
    </div>
    <div
      className="field"
    >
      <input
        className="input is-medium"
        name="email"
        onChange={undefined}
        placeholder="Enter an email address"
        required={true}
        type="email"
        value=""
      />
    </div>
    <div
      className="field"
    >
      <input
        className="input is-medium"
        name="password"
        onChange={undefined}
        placeholder="Enter a password"
        required={true}
        type="password"
        value=""
      />
    </div>
    <input
      className="button is-primary is-medium is-fullwidth"
      type="submit"
      value="Submit"
    />
  </form>
</div>
`;

exports[`\`{el.formType}\` Form renders a snapshot properly 2`] = `
<div>
  <h1
    className="title is-1"
  >
    Login
  </h1>
  <hr />
  <br />
  <form
    onSubmit={[Function]}
  >
    <div
      className="field"
    >
      <input
        className="input is-medium"
        name="email"
        onChange={undefined}
        placeholder="Enter an email address"
        required={true}
        type="email"
        value=""
      />
    </div>
    <div
      className="field"
    >
      <input
        className="input is-medium"
        name="password"
        onChange={undefined}
        placeholder="Enter a password"
        required={true}
        type="password"
        value=""
      />
    </div>
    <input
      className="button is-primary is-medium is-fullwidth"
      type="submit"
      value="Submit"
    />
  </form>
</div>
`;
